File: packages/php/src/Resources/SendingIdentifiers.php
<?php

declare(strict_types=1);

namespace MissionInbox\Resources;

use MissionInbox\Http\Client;

/**
 * The `sendingIdentifiers` resource. Access via `$mi->sendingIdentifiers`.
 *
 * A sending identifier is an approved `From:` address on the account. It is
 * not a mailbox — no inbox, no password, no IMAP.
 */
final class SendingIdentifiers
{
    public function __construct(private readonly Client $http) {}

    /**
     * List every registered sending identifier.
     *
     * @return array<array<string, mixed>>
     */
    public function list(): array
    {
        return $this->http->request('GET', '/api/sending-identifiers');
    }

    /** Retrieve a single sending identifier by UUID. */
    public function get(string $id): array
    {
        return $this->http->request('GET', '/api/sending-identifiers/' . rawurlencode($id));
    }

    /**
     * Register a new sending identifier.
     *
     * @param array{emailAddress: string, displayName?: string} $params
     */
    public function create(array $params): array
    {
        return $this->http->request('POST', '/api/sending-identifiers', body: $params);
    }

    /**
     * Update the identifier's display name.
     *
     * @param array{displayName?: string} $params
     */
    public function update(string $id, array $params): array
    {
        return $this->http->request('PATCH', '/api/sending-identifiers/' . rawurlencode($id), body: $params);
    }

    /** Delete a sending identifier. */
    public function delete(string $id): array
    {
        return $this->http->request('DELETE', '/api/sending-identifiers/' . rawurlencode($id));
    }
}

File: packages/java/src/main/java/com/missioninbox/resources/SendingIdentifiers.java
package com.missioninbox.resources;

import com.missioninbox.http.HttpClient;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The {@code sendingIdentifiers} resource. Access via {@code mi.sendingIdentifiers}.
 *
 * <p>A sending identifier is an approved {@code From:} address on the account.
 * It is not a mailbox — no inbox, no password, no IMAP.
 */
public final class SendingIdentifiers {

    private final HttpClient http;

    public SendingIdentifiers(HttpClient http) {
        this.http = http;
    }

    /** List every registered sending identifier. */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> list() {
        return (List<Map<String, Object>>) http.request(
                "GET", "/api/sending-identifiers", null, null);
    }

    /** Retrieve a single identifier by UUID. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> get(String id) {
        return (Map<String, Object>) http.request(
                "GET", "/api/sending-identifiers/" + encode(id), null, null);
    }

    /**
     * Register a new sending identifier.
     *
     * <p>Recognised keys: {@code emailAddress} (required), {@code displayName}.
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> create(Map<String, Object> params) {
        return (Map<String, Object>) http.request(
                "POST", "/api/sending-identifiers", null, params);
    }

    /** Update the identifier's display name. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> update(String id, Map<String, Object> params) {
        return (Map<String, Object>) http.request(
                "PATCH", "/api/sending-identifiers/" + encode(id), null, params);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> delete(String id) {
        return (Map<String, Object>) http.request(
                "DELETE", "/api/sending-identifiers/" + encode(id), null, null);
    }

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }
}
